Imbalanced Dataset:

A dataset with unequal class distribution

In [ ]:
#importing libraries
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('/content/credit_data.csv')

In [ ]:
df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0.0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0.0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0.0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0.0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0.0


In [ ]:
df.shape

(233743, 31)

In [ ]:
df.describe()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
count,233743.000000,233743.000000,233743.000000,233743.000000,233743.000000,233743.000000,233743.000000,233743.000000,233743.000000,233743.000000,...,233742.000000,233742.000000,233742.000000,233742.000000,233742.000000,233742.000000,233742.000000,233742.000000,233742.000000,233742.000000
mean,80744.556808,-0.060469,-0.016791,0.157748,0.039336,-0.056242,0.026770,-0.022946,0.002832,0.002144,...,-0.005171,-0.019462,-0.008981,0.000484,0.033355,0.002739,0.000013,0.001630,91.131255,0.001810
std,40413.188080,1.955791,1.671663,1.499277,1.413553,1.377785,1.324748,1.231813,1.207934,1.120755,...,0.743719,0.711183,0.639748,0.606398,0.510720,0.486524,0.402776,0.334182,252.681921,0.042502
min,0.000000,-56.407510,-72.715728,-33.680984,-5.683171,-42.147898,-26.160506,-43.557242,-73.216718,-13.434066,...,-34.830382,-10.933144,-44.807735,-2.836627,-10.295397,-2.604551,-22.565679,-11.710896,0.000000,0.000000
25%,48387.500000,-0.946223,-0.604117,-0.614238,-0.828721,-0.742629,-0.737130,-0.567271,-0.197427,-0.660264,...,-0.225536,-0.533468,-0.166555,-0.349724,-0.278070,-0.330830,-0.069938,-0.048758,6.000000,0.000000
50%,74458.000000,-0.024779,0.065290,0.347740,0.033682,-0.114316,-0.241852,0.013491,0.031993,-0.065175,...,-0.033686,-0.008582,-0.021617,0.046787,0.069209,-0.059553,0.003106,0.015371,23.450000,0.000000
75%,122275.000000,1.247569,0.780865,1.139026,0.825644,0.542342,0.429985,0.537524,0.331523,0.611521,...,0.170855,0.479004,0.129252,0.428510,0.370885,0.251679,0.089164,0.077984,79.990000,0.000000
max,147686.000000,2.454930,22.057729,9.382558,16.875344,34.801666,22.529298,36.877368,20.007208,15.594995,...,27.202839,10.503090,22.528412,4.022866,7.519589,3.517346,12.152401,33.847808,19656.530000,1.000000


In [ ]:
df.tail()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
233738,147684.0,1.938933,-0.726856,-0.197019,0.670379,-1.043038,-0.473694,-0.682871,-0.129264,-0.323037,...,-0.434638,-0.810170,0.369899,-0.119773,-0.495191,-0.930564,0.074385,-0.005836,74.00,0.0
233739,147684.0,-0.458365,-0.734428,-1.299382,-1.498641,-1.073566,0.340757,0.796631,0.184616,-0.758188,...,0.819259,2.447577,0.218824,0.838119,-1.388117,-0.133513,0.219108,0.191030,302.06,0.0
233740,147685.0,2.072601,-0.169893,-1.165646,0.150432,0.120419,-0.468132,-0.062591,-0.207555,0.631411,...,-0.042515,0.036454,0.191641,0.538720,-0.115481,0.535193,-0.068240,-0.054121,10.00,0.0
233741,147685.0,2.302987,-1.402751,-0.661621,-1.441633,-1.615912,-1.008410,-1.235227,-0.195914,-0.952930,...,-0.174727,0.010670,0.276110,0.000651,-0.281599,-0.176316,0.017253,-0.050112,5.00,0.0
233742,147686.0,1.406398,-1.009139,-3.381354,0.247401,1.245001,-0.374091,1.514391,-0.569273,-0.381729,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
df.isna().sum()

,0
Time,0
V1,0
V2,0
V3,0
V4,0
V5,0
V6,0
V7,0
V8,0
V9,0


In [ ]:
#distribution of the two classes
df['Class'].value_counts()

,count
Class,
0.0,233319
1.0,423


This is Highly imbalanced dataset.

0 -- > Legit Transactions

1 --> Fraud Transactions

In [20]:
#seperating the legit and fraud transcations
legit = df[df['Class'] == 0]
fraud = df[df['Class']== 1]

In [21]:
print(legit.shape)

(233319, 31)


In [22]:
print(fraud.shape)

(423, 31)


Under-Sampling:

we are building a sample dataset containing similar distribution of legit transactions and fraud transactions. So we have about 423 fraud and 233319 legit which is very imbalanced. we take similar amount of legit transcations, say for example 500

In [24]:
#take 500 random sample from legit data to balance
legit_sample = legit.sample(n=500)

In [26]:
legit_sample.shape

(500, 31)

In [27]:
#concatenate the legit_sample and fraud sample
new_df = pd.concat([legit_sample,fraud], axis=0)

In [28]:
new_df.shape

(923, 31)

In [29]:
new_df['Class'].value_counts()

,count
Class,
0.0,500
1.0,423
